### DISCUSSION TITLE
tutorials/navigation/navigation_using_navigationobstacles

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#131] [User-contributed notes]
# tutorials/navigation/navigation_using_navigationobstacles

2D and 3D versions of NavigationObstacles nodes are available as NavigationObstacle2D and NavigationObstacle3D respectively. Navigation obstacles are dual purpose in that they can affect both the n...

https://docs.godotengine.org/en/stable/tutorials/navigation/navigation_using_navigationobstacles.html

**Comment by ChickensioDev** [comment]
ur supposed to put the navigation obstacle 2ds as child of navigation region 2d in order for it to work. i spent over an hour searching everywhere. youtube, docs, and it wasnt mentioned and i just happened to find that out by chance.

**Reply by Dirk-Sandberg** [reply]
What if I'm using a NavigationLayer inside of a TileMapLayer node? I have a separate scene (a pillar, not part of the tileset) which I want to use as a NavigationObstacle2D, but the pillar doesn't seem to affect the NavigationLayer.

**Reply by ankhzet** [reply]
I enden up putting all layers inside a NavigationRegion2D node and re-baking nav mesh after some minimal delay after  the map node is `_ready()` and all obstacle scenes instantiated.
In my case, the obstacle scene(s) do not have NavigationObstacle2D or NavigationObstacle2D child nodes, it has a collider, which gets properly baked in the nav mesh, but when I tried NavigationObstacle2D nodes, they worked just fine.
Relevant code:

```gdscript
class_name Map extends NavigationRegion2D

var _size = Vector2(0, 0)


func load_map(prefab: PackedScene):
	var layers = prefab.instantiate().get_children()

	for child in layers:
		child.owner = null
		child.reparent(self)

		# if nav mash isn't shown in debug mode, try to disable the layers, visualizer might have low z-index
		# if child is TileMapLayer:
			# child.visible = false

	for child in get_children():
		if child is TileMapLayer:
			_size = child.get_used_rect().size * child.rendering_quadrant_size # use first layer (ground layer) to get map size
			break

	camera.adjust(_size)

# this callback is called by the child scenes of the scene layers (in the _ready())
# you can use parent reference search or a signal
func add_entity(entity: Entity, pos: Vector2) -> Entity:
	if entity.is_static():
		_debounce()

var _timer: SceneTreeTimer

func _debounce():
	var timer = get_tree().create_timer(0.1)
	_timer = timer
	await timer.timeout
	if _timer != timer:
		return

	_timer = null
	bake_navigation()

# i only bake nav mesh once on load, if obstacles are added dynamically, 
# this should be split into initial-mesh-bake and mesh-update parts
func bake_navigation():
	navigation_polygon = NavigationPolygon.new()

	var bounding_outline = PackedVector2Array([
		Vector2(0, 0),
		Vector2(0, _size.y),
		Vector2(_size.x, _size.y),
		Vector2(_size.x, 0),
	])
	navigation_polygon.agent_radius = 1
	navigation_polygon.add_outline(bounding_outline)

	self.bake_navigation_polygon()
```

**Reply by Floofwoof16** [reply]
Thank you for mentioning the need to put the obstacles as children of the region, I also didn't find that info anywhere and needed it.

**Reply by partykapartyka** [reply]
@Dirk-Sandberg - check out this https://youtu.be/7ZAF_fn3VOc?si=5ITKrjqZZq3SoOGp

I took the approach from this video and adjusted it so that you can change the navigation map during play. I'm still learning so not sure if this is the right approach but once you have a seperate TileMapLayer with some tiles which mark the place which has to be unwalkable (in my case called 'blockade'), then you just add this script to the main TileMapLayer:

```gdscript
extends TileMapLayer

@onready var blockade: TileMapLayer = $"../blockade"
@onready var blockade_on: bool = false

func _use_tile_data_runtime_update(coords: Vector2i) -> bool:
	if coords in blockade.get_used_cells():
		return true
	return false
	
func _tile_data_runtime_update(coords: Vector2i, tile_data: TileData) -> void:
	if blockade_on:
		if coords in blockade.get_used_cells():
			tile_data.set_navigation_polygon(0,null)
#
func _input(event: InputEvent) -> void:
	if event.is_action_pressed("ui_accept"):
		blockade_on = true
		notify_runtime_tile_data_update()
```
edit: Just noticed that this won't work in your case. Didn't read carefully enough the first time. Maybe if you would add the scene as part of the TileSet?

**Comment by iway1** [comment]
i find it very confusing this article is about the NavigationObstacle classes but the code examples don't use the NavigationObstacle classes
